server/rpc.go

```diff
@@ -50,6 +51,88 @@ type Broadcaster interface {
 	Sign([]byte) ([]byte, error)
 }
 
+type BroadcastSessionsManager struct {
+	broadcastSessions []*BroadcastSession
+	broadcastSessMap  map[string]*BroadcastSession
+	sessLock          *sync.Mutex
+	createSessions    func() ([]*BroadcastSession, error)
+}
+
+func NewSessionManager(node *core.LivepeerNode, pl core.PlaylistManager) *BroadcastSessionsManager {
+	bsm := &BroadcastSessionsManager{
+		broadcastSessMap: make(map[string]*BroadcastSession),
+		createSessions:   func() ([]*BroadcastSession, error) { return selectOrchestrator(node, pl) },
+		sessLock:         &sync.Mutex{},
+	}
+	bsm.refreshSessions()
+	return bsm
+}
+
+func (bsm *BroadcastSessionsManager) selectSession() *BroadcastSession {
+	// minOrchs is maximum number of inflight sessions
+	minOrchs := int(HTTPTimeout / SegLen)
+	bsm.sessLock.Lock()
+	defer bsm.sessLock.Unlock()
+	numSess := len(bsm.broadcastSessions)
+
+	if numSess < minOrchs {
+		go bsm.refreshSessions()
+	}
+
+	if numSess <= 0 {
+		return nil
+	}
+
+	last := numSess - 1
+	sess, sessions := bsm.broadcastSessions[last], bsm.broadcastSessions[:last]
+	bsm.broadcastSessions = sessions
+	return sess
+}
+
+func (bsm *BroadcastSessionsManager) removeSession(session *BroadcastSession) {
+	bsm.sessLock.Lock()
+	delete(bsm.broadcastSessMap, session.OrchestratorInfo.Transcoder)
+	bsm.sessLock.Unlock()
+}
+
+func (bsm *BroadcastSessionsManager) completeSession(sess *BroadcastSession) {
+	bsm.sessLock.Lock()
+	defer bsm.sessLock.Unlock()
+
+	if _, ok := bsm.broadcastSessMap[sess.OrchestratorInfo.Transcoder]; ok {
+		bsm.broadcastSessions = append(bsm.broadcastSessions, sess)
+	}
+}
+
+var selectOrchs = func(n *core.LivepeerNode, pl core.PlaylistManager) ([]*BroadcastSession, error) {
+	return selectOrchestrator(n, pl)
+}
+
+func (bsm *BroadcastSessionsManager) refreshSessions() {
+	newBroadcastSessions, err := bsm.createSessions()
+	if err != nil {
+		return
+	}
+
+	// include all O's being used successfully in refreshed list
```

Comment:
This comment is unclear; I think here it's just exiting early if we receive an empty list from `createSessions` ?